File: clients/js/src/features/config_plugin_package.ts
import { Input } from "../input";

import { PolywrapClientConfigBuilder, PolywrapClient } from "@polywrap/client-js";
import { PluginPackage } from "@polywrap/plugin-js";

export async function runTestCase(input: unknown): Promise<void> {
  const inputObj = Input.expectObject<{
    uri: unknown;
    method: unknown;
    args: unknown;
  }>(input);
  const uri = Input.expectUri(inputObj.uri).uri;
  const method = Input.expectString(inputObj.method);
  const args = Input.expectObject<Record<string, unknown>>(
    inputObj.args
  );

  console.log("Creating PluginPackage");

  const pluginPackage = PluginPackage.from(() => ({
    add: (args: { a: number, b: number }) => {
      return args.a + args.b;
    }
  }));

  console.log("Adding PluginPackage to ClientConfig");

  const config = new PolywrapClientConfigBuilder()
    .addPackage(uri, pluginPackage)
    .build();

  const client = new PolywrapClient(config);

  console.log("Invoking PluginPackage");

  const result = await client.invoke({
    uri,
    method,
    args
  });

  if (result.ok) {
    console.log("Success!");
  }
}

File: clients/py/src/features/config_plugin_package.py
from typing import Any, TypedDict

from polywrap import (
    PluginModule,
    PluginPackage,
    PolywrapClient,
    PolywrapClientConfigBuilder,
)
from pydantic import BaseModel, validator

from validators import UriStr, validate_uri


class Args(TypedDict):
    a: int
    b: int


class TestCaseInput(BaseModel):
    uri: UriStr
    method: str
    args: Args

    @validator("uri")
    def valid_uri(cls, v: Any):
        return validate_uri(v)


class Plugin(PluginModule[None]):
    def __init__(self, config: None):
        super().__init__(config)

    def add(self, args: Args, *_: Any):
        return args["a"] + args["b"]


def run_test_case(input: Any) -> None:
    input_obj = TestCaseInput.validate(input)

    print("Creating PluginPackage")

    plugin_package = PluginPackage(Plugin(None), NotImplemented)

    print("Adding PluginPackage to ClientConfig")

    config = (
        PolywrapClientConfigBuilder().set_package(input_obj.uri, plugin_package).build()
    )

    client = PolywrapClient(config)

    print("Invoking PluginPackage")

    try:
        client.invoke(uri=input_obj.uri, method=input_obj.method, args=input_obj.args)
        print("Success!")
    except Exception as e:
        print(e)
